Add editor setting for spin slider sensibility

winston-yallow opened the pull request:
This PR adds an editor setting "Float Drag Speed" to control the base speed of dragging EditorSpinSlider nodes. It is applied directly to `grabbing_spinner_dist_cache` which means that all the other logic (pressing shift to slow down; pressing control to snap to integers; and always being a multiple of `step`) works flawlessly with the new speed. The initial value for the setting is `5.0` to increase the base speed.

Closes godotengine/godot-proposals/issues/3017
Probably also makes godotengine/godot-proposals/issues/2760 obsolete. That proposal requests a key to increase the drag speed. With this PR it would instead have an increased base speed, using the existing shift modifier to slow the speed down if needed. This is in line with other applications like blender that use a similar approach.

winston-yallow committed:
Add editor setting for spin slider sensibility

Calinou commented:
I think the setting should influence the speed of dragging both floating-point and integer values.

winston-yallow commented:
it does already influence the integer values, as those are calculated based on this floating point value

winston-yallow commented:
Rebased on current master and adjusted to use `EDITOR_SETTING` macro like all other settings in master. Therefore this cannot be easily cherry-picked for 3.x anymore. Should I make a second PR for 3.x once this is accepted?

winston-yallow commented:
Updated to work with current master. From my side this is ready now.

winston-yallow commented:
And now the CI should be happy too :crossed_fingers:

akien-mga commented on editor/gui/editor_spin_slider.cpp:
```diff
@@ -69,6 +70,7 @@ void EditorSpinSlider::gui_input(const Ref<InputEvent> &p_event) {
 				} else {
 					grabbing_spinner_attempt = true;
 					grabbing_spinner_dist_cache = 0;
+					grabbing_spinner_speed = EditorSettings::get_singleton()->get("interface/inspector/float_drag_speed");
```
Fetching the editor setting every time there's a click event seems a bit wasteful, that's a lot of processing for something that's not going to change often.

I would move this at least to a member variable set in the constructor, so it's fetched only once per slider.

winston-yallow replied:
makes sense, done!

akien-mga commented on editor/gui/editor_spin_slider.cpp:
```diff
@@ -705,4 +706,5 @@ EditorSpinSlider::EditorSpinSlider() {
 	grabber->connect("mouse_entered", callable_mp(this, &EditorSpinSlider::_grabber_mouse_entered));
 	grabber->connect("mouse_exited", callable_mp(this, &EditorSpinSlider::_grabber_mouse_exited));
 	grabber->connect("gui_input", callable_mp(this, &EditorSpinSlider::_grabber_gui_input));
+	grabbing_spinner_speed = EditorSettings::get_singleton()->get("interface/inspector/float_drag_speed");
```
Welp, it's crashing actually here. Try to set it in `NOTIFICATION_ENTER_TREE`..

winston-yallow replied:
That's weird, it's not crashing for me. I mean I'll change it, but any idea why it crashes? Trying to understand this

akien-mga commented:
It crashes when generating the documentation, instantiating `EditorSpinSlider` while probably `EditorSettings` doesn't exist as a singleton. From the CI logs:
```
 editor/gui/editor_spin_slider.cpp:709:63: runtime error: member access within null pointer of type 'struct EditorSettings'
core/object/object.cpp:311:6: runtime error: member access within null pointer of type 'const struct Object'

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.1.dev.custom_build (67caeac3de2cb07e1907fbac65cb4e99d19a8914)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(+0x3ae94384) [0x55b998cc9384] (??:?)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x43090) [0x7f91b19f2090] (??:0)
[3] Object::get(StringName const&, bool*) const (??:?)
[4] EditorSpinSlider::EditorSpinSlider() (??:?)
[5] Object* ClassDB::creator<EditorSpinSlider>() (??:?)
[6] ClassDB::instantiate(StringName const&) (??:?)
[7] ClassDB::class_get_default_property_value(StringName const&, StringName const&, bool*) (??:?)
[8] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(+0x44e43d37) [0x55b9a2c78d37] (??:?)
[9] DocTools::generate(bool) (??:?)
[10] EditorHelp::generate_doc(bool) (??:?)
[11] EditorNode::EditorNode() (??:?)
[12] Main::start() (??:?)
[13] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(main+0x3da) [0x55b998cc8a83] (??:?)
[14] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f91b19d3083] (??:0)
[15] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(_start+0x2e) [0x55b998cc85ee] (??:?)
-- END OF BACKTRACE --
================================================================
```

akien-mga commented:
Thanks!
